### PR TITLE
🎁 Reward takes completion ids

### PR DESCRIPTION
**Pull request by qgallouedec**
This PR provides access to completion tokens in reward functions. This enables, for example, the implementation of Soft Overlong Punishment (from DAPO).

### CODE REVIEW

File: docs/source/grpo_trainer.md [hunk 1]
```diff
@@ -320,9 +321,29 @@ The [`GRPOTrainer`] supports using custom reward functions instead of dense rewa
...
+
+```python
+>>> prompts = ["The sky is", "The sun is"]  # not used in the reward function, but the trainer will pass it
+>>> completions = [" blue.", " in the sky."]  # not used in the reward function, but the trainer will pass it
+>>> completions_ids = [[6303, 13], [304, 279, 12884, 13]]
+>>> reward_func(prompts=prompts, completions=completions, completions_ids=completions_ids)
+[2.0, 4.0]
+```
+
+#### Example 1bis: Reward longer completions (based in the number of caracters)
```
Comment (lewtun) [inline docs/source/grpo_trainer.md:340]: nit:
```suggestion
#### Example 1.1: Reward longer completions (based in the number of characters)
```

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1163,9 +1169,11 @@ def _generate_and_score_completions(
                         rewards_per_func[:, i] = reward_func(**reward_inputs).logits[:, 0]  # Shape (B*G,)
                 else:
                     # Repeat all input columns (but "prompt" and "completion") to match the number of generations
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:1171]:
```suggestion
                    # Repeat all input columns (but "prompt", "completion", and "completion_ids") to match the number of generations
```

**Review by lewtun — APPROVED** [review]
LGTM!

**Review by shirinyamani — APPROVED** [review]
LGTM, I see its also well documented that its char-level not token. Great! I will close the earlier from me in favor of this then!
